File: go-leetcode/array/34.在排序数组中查找元素的第一个和最后一个位置.go
package leetcode

func searchRange(nums []int, target int) []int {
	l, r := 0, len(nums)-1

	for l <= r {
		mid := l + (r-l)/2

		if nums[mid] == target {
			i, j := 1, 1

			// 向左扩展
			for mid >= i && nums[mid-i] == target {
				i++
			}
			// 向右扩展（增加边界判断）
			for mid+j < len(nums) && nums[mid+j] == target {
				j++
			}
			return []int{mid - i + 1, mid + j - 1}
		} else if nums[mid] > target {
			r = mid - 1
		} else {
			l = mid + 1
		}
	}
	return []int{-1, -1}
}

File: python-leetcode/array/34.在排序数组中查找元素的第一个和最后一个位置.py
from typing import List


class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        left, right = 0, len(nums) - 1

        while left <= right:
            mid = left + (right - left) // 2

            if nums[mid] == target:
                i, j = 1, 1

                # 向左扩展
                while mid >= i and nums[mid - i] == target:
                    i += 1
                # 向右扩展
                while mid + j < len(nums) and nums[mid + j] == target:
                    j += 1

                return [mid - i + 1, mid + j - 1]

            elif nums[mid] > target:
                right = mid - 1
            else:
                left = mid + 1

        return [-1, -1]
